### PR TITLE
[PI-504] delete product

### PR DESCRIPTION
**Pull request by jaklinger**
## Summary

**Remove items from this list if they are not relevant. Remove this line once this has been done**

- Routine Change
- :exclamation: Breaking Change
- :robot: Operational or Infrastructure Change
- :sparkles: New Feature
- :warning: Potential issues that might be caused by this change

### Details

Add any summary information of what is in the change. **Remove this line if you have nothing to add.**

## Pull Request Naming

Pull requests should be named using the following format:

```text
Tag: [AEA-NNNN] - Short description
```

Tag can be one of:

- `Fix` - for a bug fix. (Patch release)
- `Update` - either for a backwards-compatible enhancement or for a rule change that adds reported problems. (Patch release)
- `New` - implemented a new feature. (Minor release)
- `Breaking` - for a backwards-incompatible enhancement or feature. (Major release)
- `Docs` - changes to documentation only. (Patch release)
- `Build` - changes to build process only. (No release)
- `Upgrade` - for a dependency upgrade. (Patch release)
- `Chore` - for refactoring, adding tests, etc. (anything that isn't user-facing). (Patch release)

If the current release is x.y.z then
- a patch release increases z by 1
- a minor release increases y by 1
- a major release increases x by 1

Correct tagging is necessary for our automated versioning and release process.

The description of your pull request will be used as the commit message for the merge, and also be included in the changelog. Please ensure that your title is sufficiently descriptive.

### Rerunning Checks

If you need to rename your pull request, you can restart the checks by either:

- Closing and reopening the pull request
- pushing an empty commit 
  ```bash
  git commit --allow-empty -m 'trigger build'
  git push
  ```
- Amend your last commit and force push to the branch
  ```bash
  git commit --amend --no-edit
  git push --force
  ```

Rerunning the checks from within the pull request will not use the updated title.

### CODE REVIEW

File: src/layers/domain/core/cpm_product/v1.py [hunk 1]
```diff
@@ -31,12 +34,26 @@ class CpmProductKeyAddedEvent(Event):
     product_team_id: UUID
     name: str
     ods_code: str
+    status: Status
     created_on: str
     updated_on: str = None
```
Comment (megan-bower4) [inline src/layers/domain/core/cpm_product/v1.py:39]: Will this still default to None for updated on? Or will it always get updated when a key is added?

Comment (jaklinger) [reply src/layers/domain/core/cpm_product/v1.py:39]: yes this will default to `None`

File: src/layers/domain/core/cpm_product/v1.py [hunk 2]
```diff
@@ -62,3 +80,13 @@ def add_key(self, key_type: str, key_value: str) -> CpmProductKeyAddedEvent:
...
+
+    @event
+    def delete(self):
+        deleted_on = now()
+        product_data = self._update(
+            data=dict(
+                status=Status.INACTIVE, updated_on=deleted_on, deleted_on=deleted_on
+            )
+        )
+        return CpmProductDeletedEvent(**product_data)
```
Comment (megan-bower4) [inline src/layers/domain/core/cpm_product/v1.py:92]: Where do we set the pk and sk to like P#INACTIVE? (I tried to find where we were doing it previously for devices but also got lost so this is just for my understanding aha)

File: src/api/tests/feature_tests/features/deleteCpmProduct.failure.feature [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+      | path                             | value                                                               |
+      | resourceType                     | OperationOutcome                                                    |
+      | id                               | << ignore >>                                                        |
+      | meta.profile.0                   | https://fhir.nhs.uk/StructureDefinition/NHSDigital-OperationOutcome |
+      | issue.0.severity                 | error                                                               |
+      | issue.0.code                     | processing                                                          |
+      | issue.0.details.coding.0.system  | https://fhir.nhs.uk/StructureDefinition/NHSDigital-OperationOutcome |
+      | issue.0.details.coding.0.code    | RESOURCE_NOT_FOUND                                                  |
+      | issue.0.details.coding.0.display | Resource not found                                                  |
+      | issue.0.diagnostics              | Could not find object with key '123'                                |
```
Comment (megan-bower4) [inline src/api/tests/feature_tests/features/deleteCpmProduct.failure.feature:33]: Should this not say object anymore but productTeam?

Comment (megan-bower4) [reply src/api/tests/feature_tests/features/deleteCpmProduct.failure.feature:33]: Ah never mind the error message change was in the other ticket 🤦‍♀️

Comment (jaklinger) [reply src/api/tests/feature_tests/features/deleteCpmProduct.failure.feature:33]: actually i just rebased so this has been updated now!

File: src/layers/domain/core/cpm_product/v1.py (hunk 2 above)
Comment (jaklinger) [reply src/layers/domain/core/cpm_product/v1.py:92]: ah yes very good shout

**Review by megan-bower4 — APPROVED** [review]

**Pull request merged**